building-epics/cross-compile-epics-and-a-ioc-to-an-old-x86-linux.md

```diff
@@ -0,0 +1,367 @@
+```
+
+Change the line:
+
+```
+CROSS\_COMPILER\_TARGET\_ARCHS=
+```
+
+To:
+
+```
+CROSS\_COMPILER\_TARGET\_ARCHS=linux-686
+```
+
+This tells EPICS base to be compiled both for the host system and for the linux-686 target.
+
+Save and close. Now we will create our own target configuration file, based on a existing file.
+
+```
+cd $EPICS\_BASE/configure/os
+cp CONFIG\_SITE.Common.linux-x86 CONFIG\_SITE.Common.linux-686
+```
+
+Open CONFIG\_SITE.Common.linux-686 and edit it. Comment out the line:
+
+```
+#COMMANDLINE\_LIBRARY = READLINE
+```
+
+And uncomment:
+
+```
+COMMANDLINE\_LIBRARY = READLINE\_NCURSES
+```
+
+At the end of the file, add the lines:
+
+```
+GNU\_DIR=$HOME/x-tools/i686-unknown-linux-gnu
+GNU\_TARGET=i686-unknown-linux-gnu
+```
+
+This tells EPICS to search for both readline and ncurses libraries during compilation. The last two lines indicate the location of the toolchain and its prefix. Save and close. Last file to edit: CONFIG.Common.linux-686
+
+Change the line that says
+
+```
+VALID\_BUILDS = Ioc
+```
+
+To
+
+```
+VALID\_BUILDS = Host Ioc
+```
+
+This is needed in order to get caRepeater compiled, according to [this source](https://epics.anl.gov/tech-talk/2012/msg01102.php).
+
+### Recompile EPICS base
+
+Now, we recompile EPICS base:
+
+```
+make clean uninstall
+make
+```
+
+Hopefully, we have everything in place to start developing our IOC’s.
+
+## Example IOC
+
+Let’s create a working directory for our programs. I decided to call it ‘apps’:
+
+```
+mkdir ~/apps
+```
+
+### Creating
+
+To create an example IOC, we use a Perl script provided by EPICS:
+
+```
+cd ~/apps
+mkdir myexample
+cd myexample
+makeBaseApp.pl -t example myexample
+```
+
+The last command tells the script to create an application named ‘myexample’ using the template (option -t) ‘example’. Now we make our IOC bootable
+
+```
+makeBaseApp.pl -i -t example myexample
+```
+
+It will ask you what the target is. We went to all this trouble to be able to select:
+
+```
+linux-686
+```
+
+For the Application Name, just hit enter.
+
+### Configuring
+
+Add this line to the file ~/apps/myexample/configure/CONFIG\_SITE
+
+```
+STATIC\_BUILD=YES
+```
+
+This will statically link EPICS libraries into our executable.
+
+Now, let’s consider that you will put your IOC in the folder /home of your target system. Edit the file ~/apps/myexample/iocBoot/iocmyexample/envPaths, so it will be:
+
+```
+epicsEnvSet("ARCH","linux-686")
+epicsEnvSet("IOC","iocskel")
+epicsEnvSet("TOP","/home/myexample")
+epicsEnvSet("EPICS\_BASE","/home/myexample")
+```
+
+Note that we set EPICS base to coincide with the <top> folder of our IOC. I did it because the IOC depends on the caRepeater program, which would be present in an EPICS base if we had one in our target. Because we don’t, I’ll simply copy the caRepeater generated by the host to the the /bin folder of our IOC <top> folder:
+
+```
+cp $EPICS\_BASE/bin/linux-686/caRepeater ~/apps/myexample/bin/linux-686/
```

Comment:
escapes (where on earth did these all come from?)

Reply:
sorry :D I think is because I used a site to convert html to md, and it make these all escapes. I will double check on next PRs.